micromanager_gui/main_window.py

```diff
@@ -148,92 +159,121 @@ def add_frame_explorer(self, name, array):
         except KeyError:
             self.viewer.add_image(array, name=layer_name)
 
-    # TO DO: add the file name form the save box
-    def _on_mda_frame(self, image, event):
 
-        sequence = event.sequence
+    # create temp folder when mda starts
+    def _on_mda_started_temp_folder(self, sequence: useq.MDASequence):
 
-        event.index.setdefault('t', 0)
-        event.index.setdefault('z', 0)
+        self.temp_folder = tempfile.TemporaryDirectory(None, str(sequence.uid))
+        return self.temp_folder.name
 
-        t_stack_length = len(sequence.time_plan) or 1
-        p_stack_length = len(sequence.stage_positions)
-        z_stack_length = len(sequence.z_plan) or 1
-        c_stack_length = len(sequence.channels)
+    #get the name to use to save the active ayer
+    def get_filename(self, fname, list_dir):
+        val = int(fname.split('_')[-1])
+        while True:
+            new_val = '{0:03}'.format(val)
+            fname = fname[:-3] + new_val
+            if ((fname + '.tif') or (fname + '.tiff')) not in set(list_dir):
+                break
+            else:
+                val += 1
+        return fname
 
-        event_index_t = event.index["t"]
-        event_index_p = event.index["p"]
-        event_index_z = event.index["z"]
-        event_index_c = event.index["c"]
-        
-        file_name = 'mda'
+    # delete temp folder and files when mda is done and save layer
+    # and add increment to save name
+    def _on_mda_finished_save_layer(self, sequence: useq.MDASequence):
+            
+        if self.mda.save_groupBox.isChecked():
 
-        layer_name = (
-            f'{file_name}_[{sequence.axis_order}]_p{p_stack_length}_'
-            f't{t_stack_length}_z{z_stack_length}_c{c_stack_length}'
-        )
+            active_layer = self.viewer.layers.selection.active.name
 
-        try:
-            layer = self.viewer.layers[layer_name]
+            fname =  self.mda.fname_lineEdit.text()
 
-            if sequence.axis_order == 'tpzc' or sequence.axis_order == 'ptzc':            
-                #channels
-                if event_index_c > 0 and event_index_z == 0 and event_index_p == 0 and event_index_t == 0:
-                    empty_im = np.empty(((1,)*len(sequence.axis_order) + image.shape), dtype=np.uint16) 
-                    layer.data = np.concatenate((layer.data,empty_im), axis=-3)
+            save_path = Path(self.mda.dir_lineEdit.text())
+
+            list_dir = [str(i).split('/')[-1] for i in save_path.iterdir() \
```

Comment:
I recommend looking through the [pathlib](https://docs.python.org/3/library/pathlib.html) docs... They're very useulf.  Here it looks like you're looking for [glob](https://docs.python.org/3/library/pathlib.html#pathlib.Path.glob): `save_path.glob('*.tif')` ... and possibly also [Path.name](https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.name): 
```
[p.name for p in save_path.glob("*.tif")]
```

however, i think the whole `get_filename` thing probably should be reworked

Reply:
This is basically what I wanted to achieve, I've added few comments.
I'm sure there is a more correct way to do it, and I actually wanted to discuss with you on how you would approach this task :). It does what I want but I think I probably have to start looking at stuff in a different way....

```
"""
               Ckeck the input filename and modify it to add _nnn in the end.
               The get_filename(fname,list_dir) function check if the name 
               is already present in the save_path and incremant the _nnn accordingly.

               filename examples: user input -> output:
                - mda       -> mda_000
                - mda_3     -> mda_003
                - mda_0001  -> mda_001
                - mda1      -> mda1_000
            """
            try:
                n = fname.split('_')[-1]
                
                if len(n) == 3 and int(n)>= 0:
                    fname = self.get_filename(fname,list_dir)
                
                elif len(n) != 3 and len(n) <=4 and int(n) >= 0:
                    s = ''
                    for i in fname.split('_')[:-1]:
                        s = s + i + '_'
                        print(s)
                    print('s', s)
                    fname = s + '{0:03}'.format(int(n))
                    fname = self.get_filename(fname,list_dir)
                
                else:
                    fname = fname + '_000'
                    fname = self.get_filename(fname,list_dir)

            except ValueError:
                fname = fname + '_000'
                fname = self.get_filename(fname,list_dir)
```

This function instead just checks if the filename is already present in the save_path folder and change the _nnn in the end of the filename accordingly:

```
def get_filename(self, fname, list_dir):
        """
            check if the filename_nnn used to save the layer exists
            and increment _nnn accordingly.
        """
        val = int(fname.split('_')[-1])
        while True:
            new_val = '{0:03}'.format(val)
            fname = fname[:-3] + new_val
            if ((fname + '.tif') or (fname + '.tiff')) not in set(list_dir):
                break
            else:
                val += 1
        return fname
```